### DISCUSSION TITLE
Automatic splitting and optimization of periodic complex queries

### DISCUSSION THREAD

**Discussion by HEEE61** [openGemini/openGemini#391] [Ideas]
We have learned that currently, in the context of periodic complex queries, OpenGemini uses manual splitting to break down query tasks into multiple executable subtasks. There are still areas that can be optimized. Firstly, manual splitting requires a significant amount of human effort, and secondly, the selected time windows and labels after splitting may not necessarily be the optimal combination.

Therefore, we propose a new query splitting solution. This solution is designed to automatically split incoming query tasks while simultaneously selecting the best combination for each subtask to achieve the optimal balance between resource consumption and query efficiency. The flowchart for this solution is as follows:

![流程图](https://github.com/HEEE61/Pictures/blob/main/flowchart.png?raw=true)

This solution proposes a framework for splitting periodic queries, which includes two modules: coarse-grained optimization search and fine-grained optimization search. These modules are responsible for query decomposition and aggregation granularity selection, respectively.

The coarse-grained search utilizes equivalent optimization rules to construct a search tree with query plans as nodes and query plan optimization rules as edges. The root node represents the original query plan, and each node logically represents a query plan that divides the complete business into multiple stream aggregation and CQ tasks. It employs a neural network model to estimate the query cost for these query plans and uses the Monte Carlo tree search method to iteratively expand and select nodes in the search tree, ultimately producing a query plan to serve as input for the fine-grained optimization search.The following image represents the structure of the coarse-grained search:

![粗粒度搜索结构](https://github.com/HEEE61/Pictures/blob/main/WriteUse/FirStructure.png?raw=true)

The fine-grained search selects the time windows and labels required for aggregation and grouping in stream aggregation and CQ tasks. Each selection is represented by a node. Because different selections lead to multiple subsequent nodes for each current node, the search space is structured as a tree. We employ various search algorithms to explore this search space, comparing possible solutions based on defined cost evaluation criteria. Ultimately, the best combination of time windows and labels for stream aggregation and CQ tasks is determined and outputted.The following image represents the search space established for fine-grained search:

![细粒度搜索空间](https://github.com/HEEE61/Pictures/blob/main/WriteUse/SecStructure.png?raw=true)

For the optimization process involving two different granularities, we propose three different combination methods:

1. Fine-Grained Search at Every Step: In this approach, fine-grained search is performed at each optimization node in the process.

2. Fine-Grained Search on the Output of Coarse-Grained Optimization: Here, fine-grained search is applied to the output results of the coarse-grained optimization.

3. Rule-Based Selection for Fine-Grained Search: This method involves using predefined rules to determine whether to enter the fine-grained search based on certain criteria or conditions.

These three methods provide different ways to integrate coarse-grained and fine-grained optimization processes to achieve the final optimization results.

Finally, In the side of comparing the optimal solutions, to accurately measure the efficiency and resource utilization of each possible solution and evaluate the quality of the decomposition results, you can specify cost evaluation criteria through two approaches: metric mapping and benchmark testing.

Metric Mapping: This approach involves defining specific metrics or criteria that are relevant to the efficiency and resource utilization of the solutions. These metrics can include factors such as query execution time, system resource consumption, or any other relevant performance indicators. The solutions are then evaluated based on how well they meet these defined metrics.

Benchmark Testing: Benchmark testing involves running the solutions in a controlled environment or against predefined test cases to measure their performance. This approach allows you to compare the actual performance of each solution under specific conditions and provides empirical data on their efficiency.

![代价公式与指标映射](https://github.com/HEEE61/Pictures/blob/main/WriteUse/Cost.png?raw=true)

We conducted write and query tests on openGemini, established an initial cost formula, and determined the weights of its individual components. During actual operation, we periodically adjust the weights of the cost formula's components by monitoring metric changes to achieve the highest level of accuracy for the cost formula.

The above is the main content of our plan.

**Comment by shilinlee** [comment]
That‘s a very nice good idea. This can greatly promote the utilization of CQ and allow users to focus more on the business.
